src/kernel/sporadic.c

```diff
@@ -247,46 +248,83 @@ static bool_t refill_head_overlapping(sched_context_t *sc)
     }
 }
 
-void refill_budget_check(ticks_t usage)
+static inline bool_t head_refill_overrun(sched_context_t *sc, ticks_t usage)
 {
-    sched_context_t *sc = NODE_STATE(ksCurSC);
-    assert(!isRoundRobin(sc));
-    REFILL_SANITY_START(sc);
+    refill_t head = *refill_head(sc);
+    return head.rAmount <= usage && head.rTime < MAX_RELEASE_TIME;
+}
+
+static inline ticks_t charge_entire_head_refill(sched_context_t *sc, ticks_t usage)
+{
+    refill_t head = *refill_head(sc);
+
+    if (refill_single(sc)) {
+        refill_head(sc)->rTime = head.rTime + sc->scPeriod;
+    } else {
+        refill_t old_head = refill_pop_head(sc);
+        old_head.rTime += sc->scPeriod;
+        schedule_used(sc, old_head);
+    }
+
+    return usage - head.rAmount;
+}
+
+static inline ticks_t handle_overrun(sched_context_t *sc, ticks_t usage)
+{
+    ticks_t current_usage = usage;
 
     /*
-     * We charge entire refills in a loop until we end up with a partial
-     * refill or at a point where we can't place refills into the future
-     * without integer overflow.
+     * We repeatedly charge the head refill in its entirety until we end up
+     * with a partial refill, or at a point where we can't place refills
+     * into the future without integer overflow.
      *
      * Verification actually requires that the current time is at least
      * 3 * MAX_PERIOD from the INT64_MAX value, so to ease relation to
-     * that assertion we ensure that we never delate a refill past this
+     * that assertion we ensure that we never delay a refill past this
      * point in the future.
      */
-    while (refill_head(sc)->rAmount <= usage && refill_head(sc)->rTime < MAX_RELEASE_TIME) {
-        usage -= refill_head(sc)->rAmount;
-
-        if (refill_single(sc)) {
-            refill_head(sc)->rTime += sc->scPeriod;
-        } else {
-            refill_t old_head = refill_pop_head(sc);
-            old_head.rTime += sc->scPeriod;
-            schedule_used(sc, old_head);
-        }
+    while (head_refill_overrun(sc, current_usage)) {
+        current_usage = charge_entire_head_refill(sc, current_usage);
     }
 
+    return current_usage;
+}
+
+static inline bool_t head_refill_insufficient(sched_context_t *sc)
+{
+    return refill_head(sc)->rAmount < MIN_BUDGET;
+}
+
+static inline void merge_nonoverlapping_head_refill(sched_context_t *sc)
+{
+    refill_t head = refill_pop_head(sc);
+    refill_head(sc)->rAmount += head.rAmount;
+    /* Delay head to ensure the subsequent refill doesn't end any
+     * later (rather than simply combining refills). */
+    refill_head(sc)->rTime -= head.rAmount;
+}
+
+void refill_budget_check(ticks_t usage)
+{
+    sched_context_t *sc = NODE_STATE(ksCurSC);
+    assert(!isRoundRobin(sc));
+    REFILL_SANITY_START(sc);
```

Comment:
out of curiosity, what does this do? is there a corresponding `REFILL_SANITY_START`?

Reply:
There's a REFILL_SANITY_END at the bottom. It just checks that the refill_sum didn't change and that refills are ordered and the SC is still round-robin.